### ISSUE TITLE
Dotnet stopped working on Arm64 based Rockchip RK3566

### ISSUE THREAD

**Issue by v4Lo** [dotnet/runtime#82672]
### Description

Dotnet doesn't work since version 7.0 on Linux on an Arm64 based Rockchip RK3566 (Pine64 Quartz64) computer. Any version 6.0 works without issues. No other precompiled binaries I am running have this issue.

I originally thought this was a problem with the software I was running, but it seems this is caused by changes to Dotnet. (https://github.com/TechnitiumSoftware/DnsServer/issues/521)

Running dotnet --info with the SDK installed on Dotnet 7.0 on this hardware outputs (just running an application does the same)

```
Failed to create CoreCLR, HRESULT: 0x80004005
```

It doesn't seem to be fixed in 8.0-preview1. There is different behavior though.

```
./dotnet  --info
Segmentation fault
```

```
uname -a
Linux storagepi 6.1.0-rc1 #1 SMP PREEMPT Wed Nov  9 01:08:14 UTC 2022 aarch64 GNU/Linux
```

### Reproduction Steps

1. Install Linux on a Rockchip RK3566 based computer, possibly Pine64 Quartz64 specific, no other samples available to me
2. Install dotnet 7.0 runtime and SDK
3. Run dotnet --info

### Expected behavior

No "Failed to create CoreCLR, HRESULT: 0x80004005" using dotnet 7.0, no segfault using dotnet 8.0. Instead run the application as normal.

### Actual behavior

No dotnet application works starting version 7.0 on arm64 based Pine64 Quartz64.

### Regression?

Any version 6.0 dotnet works flawlessly.

### Known Workarounds

None.

### Configuration

Debian and DietPi kernel/bootloader image
Arm64 Rockchip RK3566 based Pine64 Quartz64

**Comment by EgorBo** [comment]
Can you try with `DOTNET_EnableWriteXorExecute=0` env variable defined?

**Comment by jeffschwMSFT** [comment]
Another thing to check is your glibc version.  .NET 7 Arm64 increased the glibc version which can cause the runtime to fail to start.  An upcoming update to .NET 7 is moving the glibc version back to the same level as .NET 6.

**Comment by v4Lo** [comment]
> DOTNET_EnableWriteXorExecute=0

This does indeed fix the issue for "dotnet --info" for 7.0 aswell as 8.0, despite one giving the CoreCLR error and the other just segfaulting originally. The application I tried to run originally also works, using 7.0.

> Another thing to check is your glibc version. .NET 7 Arm64 increased the glibc version which can cause the runtime to fail to start. An upcoming update to .NET 7 is moving the glibc version back to the same level as .NET 6.

```
ldd --version
ldd (Debian GLIBC 2.31-13+deb11u5) 2.31
```

**Comment by EgorBo** [comment]
> This does indeed fix the issue

cc @janvorli on W^X

**Comment by janvorli** [comment]
@v4Lo it seems this is specific to the Quartz64 / OS, as I am running .NET 7/8 stuff on other similar Arm64 devices (Odroid N2, RPi 4) without problems. Since I don't have access to a Quartz64 device, would you be able to share a core dump for the crashing case? Here are the steps to get one that would contain enough details to be actionable:
```sh
echo 0x3f > /proc/self/coredump_filter
echo "dotnet.core" | sudo tee /proc/sys/kernel/core_pattern
dotnet --info
```

It should create a core dump named  dotnet.core in the current directory. I'd recommend gzipping it. Then sharing it via OneDrive or some other file service would be great. 

Please also share the output of
```sh
cat /etc/os-release
```

If you don't want to share it publicly, please use the email from my github profile.

**Comment by v4Lo** [comment]
> ```shell
> echo 0x3f > /proc/self/coredump_filter
> echo "dotnet.core" | sudo tee /proc/sys/kernel/core_pattern
> dotnet --info
> ```

I was unable to produce a core dump using these instructions for dotnet 7.0, no matter what I tried. No file was created in the cwd, location of the dotnet binary or /tmp. The segfaulting dotnet 8.0 produced one in /tmp.

[coredump.2654805.gz](https://github.com/dotnet/runtime/files/10852079/coredump.2654805.gz)

> 
> ```shell
> cat /etc/os-release
> ```

```
cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

Nothing terribly interesting there. It's just Debian with DietPi kernel and bootloader.

**Comment by janvorli** [comment]
I am sorry, I've forgotten to mention one more command needed to enable the dumps - the `ulimit -c unlimited` is another thing that needs to be run before running the `dotnet --info`. 
But let me take a look at the dump you've shared, maybe that would shed a light on the problem.

**Comment by v4Lo** [comment]
> ```shell
> echo 0x3f > /proc/self/coredump_filter
> echo "dotnet.core" | sudo tee /proc/sys/kernel/core_pattern
> ulimit -c unlimited
> dotnet --info
> ```

No changes, for dotnet 7.0. Still no dump. Not for dotnet --info, nor the application I'm trying to run. Dotnet 8.0 produces a much bigger dump now at the cwd.
[dotnet.core.gz](https://github.com/dotnet/runtime/files/10853238/dotnet.core.gz)

**Comment by janvorli** [comment]
Thank you, I am looking at it now.

**Comment by janvorli** [comment]
It crashes because a data structure that is being built becomes somehow corrupted in a strange way. The memory starting at a page boundary is filled by 64 bit constant 0x0000097400000000, then another block of 0x0000097400000008, then a block of 0x0000097400000010 and it goes on like this, the constant changing even in the upper 32 bits a bit. This goes on for almost 20kB. I have no idea where this comes from.

I have another question - what is the memory page size configured by your kernel? You can get it by running the following from the shell:
```sh
getconf PAGE_SIZE
```

I have tried to repro the problem with Debian 11 docker image on my Odroid N2, but it just worked.

**Comment by janvorli** [comment]
It might also be related to the brand new kernel version that your device is using. It was released only a month ago.

**Comment by v4Lo** [comment]
> ```shell
> getconf PAGE_SIZE
> 4096
> ```

I can try getting a different kernel, I'll look into how DietPi builds that image.

**Comment by v4Lo** [comment]
I switched the kernel to the latest linux 6.1 release and the issue just disappeared. I still don't understand how the kernel could affect dotnet but not anything else. The configs of the kernels are not the same so I single directionally diffed the configs. Are any of these options necessary for dotnet 7.0+ to work and dotnet 6.0 can do without?

Old config: https://gitlab.com/pgwipeout/quartz64_ci/-/raw/main/quartz64_defconfig
[quartz64_6.1.14_defconfig.txt](https://github.com/dotnet/runtime/files/10873490/quartz64_6.1.14_defconfig.txt)


```code
CONFIG_BLK_DEV_LOOP=m
CONFIG_BLK_DEV_RAM=n
CONFIG_BPF_SYSCALL=y
CONFIG_BRIDGE_NETFILTER=m
CONFIG_CAN_GS_USB=m
CONFIG_CFS_BANDWIDTH=y
CONFIG_CGROUP_BPF=y
CONFIG_IPV6=y
CONFIG_ISCSI_TCP=m
CONFIG_LEGACY_PTYS=n
CONFIG_MODULE_COMPRESS_XZ=y
CONFIG_NETFILTER_ADVANCED=y
CONFIG_NETFILTER_CONNCOUNT=m
CONFIG_NETFILTER_FAMILY_ARP=y
CONFIG_NETFILTER_FAMILY_BRIDGE=y
CONFIG_NETFILTER_NETLINK_ACCT=m
CONFIG_NETFILTER_NETLINK_GLUE_CT=y
CONFIG_NETFILTER_NETLINK_LOG=m
CONFIG_NETFILTER_NETLINK=m
CONFIG_NETFILTER_NETLINK_OSF=m
CONFIG_NETFILTER_NETLINK_QUEUE=m
CONFIG_NETFILTER_SYNPROXY=m
CONFIG_NETFILTER_XTABLES=m
CONFIG_NETFILTER_XT_CONNMARK=m
CONFIG_NETFILTER_XT_MARK=m
CONFIG_NETFILTER_XT_MATCH_BPF=m
CONFIG_NETFILTER_XT_MATCH_CGROUP=m
CONFIG_NETFILTER_XT_MATCH_CLUSTER=m
CONFIG_NETFILTER_XT_MATCH_COMMENT=m
CONFIG_NETFILTER_XT_MATCH_CONNBYTES=m
CONFIG_NETFILTER_XT_MATCH_CONNLABEL=m
CONFIG_NETFILTER_XT_MATCH_CONNLIMIT=m
CONFIG_NETFILTER_XT_MATCH_CONNMARK=m
CONFIG_NETFILTER_XT_MATCH_CPU=m
CONFIG_NETFILTER_XT_MATCH_DCCP=m
CONFIG_NETFILTER_XT_MATCH_DEVGROUP=m
CONFIG_NETFILTER_XT_MATCH_DSCP=m
CONFIG_NETFILTER_XT_MATCH_ECN=m
CONFIG_NETFILTER_XT_MATCH_ESP=m
CONFIG_NETFILTER_XT_MATCH_HASHLIMIT=m
CONFIG_NETFILTER_XT_MATCH_HELPER=m
CONFIG_NETFILTER_XT_MATCH_HL=m
CONFIG_NETFILTER_XT_MATCH_IPCOMP=m
CONFIG_NETFILTER_XT_MATCH_IPRANGE=m
CONFIG_NETFILTER_XT_MATCH_IPVS=m
CONFIG_NETFILTER_XT_MATCH_L2TP=m
CONFIG_NETFILTER_XT_MATCH_LENGTH=m
CONFIG_NETFILTER_XT_MATCH_LIMIT=m
CONFIG_NETFILTER_XT_MATCH_MAC=m
CONFIG_NETFILTER_XT_MATCH_MARK=m
CONFIG_NETFILTER_XT_MATCH_MULTIPORT=m
CONFIG_NETFILTER_XT_MATCH_NFACCT=m
CONFIG_NETFILTER_XT_MATCH_OSF=m
CONFIG_NETFILTER_XT_MATCH_OWNER=m
CONFIG_NETFILTER_XT_MATCH_PHYSDEV=m
CONFIG_NETFILTER_XT_MATCH_PKTTYPE=m
CONFIG_NETFILTER_XT_MATCH_POLICY=m
CONFIG_NETFILTER_XT_MATCH_QUOTA=m
CONFIG_NETFILTER_XT_MATCH_RATEEST=m
CONFIG_NETFILTER_XT_MATCH_REALM=m
CONFIG_NETFILTER_XT_MATCH_RECENT=m
CONFIG_NETFILTER_XT_MATCH_SCTP=m
CONFIG_NETFILTER_XT_MATCH_SOCKET=m
CONFIG_NETFILTER_XT_MATCH_STATE=m
CONFIG_NETFILTER_XT_MATCH_STATISTIC=m
CONFIG_NETFILTER_XT_MATCH_STRING=m
CONFIG_NETFILTER_XT_MATCH_TCPMSS=m
CONFIG_NETFILTER_XT_MATCH_TIME=m
CONFIG_NETFILTER_XT_MATCH_U32=m
CONFIG_NETFILTER_XT_NAT=m
CONFIG_NETFILTER_XT_SET=m
CONFIG_NETFILTER_XT_TARGET_AUDIT=m
CONFIG_NETFILTER_XT_TARGET_CLASSIFY=m
CONFIG_NETFILTER_XT_TARGET_CONNMARK=m
CONFIG_NETFILTER_XT_TARGET_CONNSECMARK=m
CONFIG_NETFILTER_XT_TARGET_CT=m
CONFIG_NETFILTER_XT_TARGET_DSCP=m
CONFIG_NETFILTER_XT_TARGET_HL=m
CONFIG_NETFILTER_XT_TARGET_HMARK=m
CONFIG_NETFILTER_XT_TARGET_IDLETIMER=m
CONFIG_NETFILTER_XT_TARGET_LED=m
CONFIG_NETFILTER_XT_TARGET_MARK=m
CONFIG_NETFILTER_XT_TARGET_MASQUERADE=m
CONFIG_NETFILTER_XT_TARGET_NETMAP=m
CONFIG_NETFILTER_XT_TARGET_NFLOG=m
CONFIG_NETFILTER_XT_TARGET_NFQUEUE=m
CONFIG_NETFILTER_XT_TARGET_RATEEST=m
CONFIG_NETFILTER_XT_TARGET_REDIRECT=m
CONFIG_NETFILTER_XT_TARGET_SECMARK=m
CONFIG_NETFILTER_XT_TARGET_TCPMSS=m
CONFIG_NETFILTER_XT_TARGET_TCPOPTSTRIP=m
CONFIG_NETFILTER_XT_TARGET_TEE=m
CONFIG_NETFILTER_XT_TARGET_TPROXY=m
CONFIG_NETFILTER_XT_TARGET_TRACE=m
CONFIG_NF_CONNTRACK_AMANDA=m
CONFIG_NF_CONNTRACK_BRIDGE=m
CONFIG_NF_CONNTRACK_BROADCAST=m
CONFIG_NF_CONNTRACK_FTP=m
CONFIG_NF_CONNTRACK_H323=m
CONFIG_NF_CONNTRACK_IRC=m
CONFIG_NF_CONNTRACK_LABELS=y
CONFIG_NF_CONNTRACK_MARK=y
CONFIG_NF_CONNTRACK_NETBIOS_NS=m
CONFIG_NF_CONNTRACK_PPTP=m
CONFIG_NF_CONNTRACK_SANE=m
CONFIG_NF_CONNTRACK_SECMARK=y
CONFIG_NF_CONNTRACK_SIP=m
CONFIG_NF_CONNTRACK_SNMP=m
CONFIG_NF_CONNTRACK_TFTP=m
CONFIG_NF_CONNTRACK_TIMEOUT=y
CONFIG_NF_CONNTRACK_TIMESTAMP=y
CONFIG_NF_CONNTRACK_ZONES=y
CONFIG_NF_CT_NETLINK_HELPER=m
CONFIG_NF_CT_NETLINK=m
CONFIG_NF_CT_NETLINK_TIMEOUT=m
CONFIG_NF_CT_PROTO_DCCP=y
CONFIG_NF_CT_PROTO_GRE=y
CONFIG_NF_CT_PROTO_SCTP=y
CONFIG_NF_CT_PROTO_UDPLITE=y
CONFIG_NF_DEFRAG_IPV4=m
CONFIG_NF_DEFRAG_IPV6=m
CONFIG_NF_DUP_IPV4=m
CONFIG_NF_DUP_IPV6=m
CONFIG_NF_DUP_NETDEV=m
CONFIG_NF_FLOW_TABLE_INET=m
CONFIG_NF_FLOW_TABLE=m
CONFIG_NF_LOG_ARP=m
CONFIG_NF_LOG_IPV6=m
CONFIG_NF_LOG_SYSLOG=m
CONFIG_NF_NAT_AMANDA=m
CONFIG_NF_NAT_FTP=m
CONFIG_NF_NAT_H323=m
CONFIG_NF_NAT_IRC=m
CONFIG_NF_NAT=m
CONFIG_NF_NAT_MASQUERADE=y
CONFIG_NF_NAT_PPTP=m
CONFIG_NF_NAT_REDIRECT=y
CONFIG_NF_NAT_SIP=m
CONFIG_NF_NAT_SNMP_BASIC=m
CONFIG_NF_NAT_TFTP=m
CONFIG_NF_REJECT_IPV4=m
CONFIG_NF_REJECT_IPV6=m
CONFIG_NF_SOCKET_IPV4=m
CONFIG_NF_SOCKET_IPV6=m
CONFIG_NF_TABLES_ARP=y
CONFIG_NF_TABLES_BRIDGE=m
CONFIG_NF_TABLES_INET=y
CONFIG_NF_TABLES_IPV4=y
CONFIG_NF_TABLES_IPV6=y
CONFIG_NF_TABLES=m
CONFIG_NF_TABLES_NETDEV=y
CONFIG_NF_TPROXY_IPV4=m
CONFIG_NF_TPROXY_IPV6=m
CONFIG_SMB_SERVER=m
CONFIG_SQUASHFS_DECOMP_MULTI_PERCPU=y
CONFIG_SQUASHFS_FILE_DIRECT=y
CONFIG_SQUASHFS_LZ4=y
CONFIG_SQUASHFS_LZO=y
CONFIG_SQUASHFS=m
CONFIG_SQUASHFS_XATTR=y
CONFIG_SQUASHFS_XZ=y
CONFIG_SQUASHFS_ZLIB=y
CONFIG_SQUASHFS_ZSTD=y
CONFIG_THERMAL_EMULATION=n
CONFIG_TUN=m
CONFIG_VIRTIO_BALLOON=m
CONFIG_VIRTIO_CONSOLE=m
CONFIG_VIRTIO_INPUT=m
CONFIG_VIRTIO_MMIO=m
CONFIG_VIRTIO_NET=m
CONFIG_VIRTIO_PCI=m
CONFIG_VXLAN=m
# CONFIG_WIREGUARD_DEBUG is not set
CONFIG_WIREGUARD=m
CONFIG_ZRAM=m
```


```
uname -a
Linux storagepi 6.1.14 #1 SMP PREEMPT Tue Feb 28 13:59:13 UTC 2023 aarch64 GNU/Linux
```

**Comment by janvorli** [comment]
The key is the W^X that was not enabled in .NET 6 by default. My guess is that rather than a config difference, there could have been a bug in the RC1 version of 6.1 kernel that you were using before that got fixed later.
W^X uses memfd_create to create an anonymous shared memory that it uses for double mapping of executable pages as writeable when we need to generate new / modify existing code. Maybe there was some issue in this area.

**Comment by v4Lo** [comment]
Alright, as no current stable release seems to be affected the issue can be closed. Thank you very much.

**Comment by v4Lo** [comment]
Issue was related to non-stable kernel version and seems to be fixed in latest 6.1
